Chain-of-Verification - Prompt Engineering

Ankush-lastmile opened the pull request:
Chain-of-Verification - Prompt Engineering

Chain-of-Verification is a prompt engineering technique to reduce hallucinations from LLMs!

Research Paper (Meta AI): https://arxiv.org/pdf/2309.11495.pdf

This recipe uses AIConfig - an open-source config-based framework for building generative AI applications.

https://github.com/lastmile-ai/aiconfig/tree/main

Ankush-lastmile commented:
[@MKhalusova](https://github.com/MKhalusova)

MKhalusova commented on notebooks/en/chain_of_verification.ipynb:
```diff
@@ -0,0 +1,470 @@
+{
```
<p>There's no need for a link to a colab notebook. Once the notebook published, a link to Colab will be auto-generated. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/29/discussion/#comment-1494821402'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Ankush-lastmile replied:
<p>removed colab link</p>

MKhalusova commented on notebooks/en/chain_of_verification.ipynb:
```diff
@@ -0,0 +1,470 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #2.    import openai
</pre><p>As this is an open source ai cookbook, let's illustrate this technique with an <strong>open source</strong> model. From the readme, I can see that these models should be supported:</p><ul><li><a href="https://huggingface.co/mistralai/Mixtral-8x7B-Instruct-v0.1" target="_blank">https://huggingface.co/mistralai/Mixtral-8x7B-Instruct-v0.1</a></li><li><a href="https://huggingface.co/chat?model=meta-llama%2FLlama-2-70b-chat-hf" target="_blank">https://huggingface.co/chat?model=meta-llama/Llama-2-70b-chat-hf</a></li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/29/discussion/#comment-1494821406'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Ankush-lastmile replied:
I've updated the model usage to use The [Mixtral-8x7B-Instruct-v0.1](https://huggingface.co/mistralai/Mixtral-8x7B-Instruct-v0.1), thanks for the suggestion!

MKhalusova commented on notebooks/en/chain_of_verification.ipynb:
```diff
@@ -0,0 +1,470 @@
+{
```
<p>It would be helpful to explain the config a little bit here. At least its essential components. Maybe add some advice on the structure/content of the prompts?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/29/discussion/#comment-1494821409'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Ankush-lastmile replied:
Added a description and links

MKhalusova commented on notebooks/en/chain_of_verification.ipynb:
```diff
@@ -0,0 +1,470 @@
+{
```
<p>The prompt here differs from the one defined in the below code cell - is this a mistake? If this is intentional, can you explain why it differs? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/29/discussion/#comment-1494821411'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Ankush-lastmile replied:
I have updated the prompt to match the below cell. This should remove any cause for confusion

MKhalusova commented on notebooks/en/chain_of_verification.ipynb:
```diff
@@ -0,0 +1,470 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #1.    # &lt;&gt;: Update baseline_prompt but ensure it is structured in a way that outputs a list of entities where each can be verified.
</pre><p>Is this TODO still needed here?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/29/discussion/#comment-1494821415'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Ankush-lastmile replied:
removed the todos, sorry!

MKhalusova commented on notebooks/en/chain_of_verification.ipynb:
```diff
@@ -0,0 +1,470 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #1.    #  &lt;&gt;: Update verification question that takes in entity as a parameter
</pre><p>Another TODO - is it needed here?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/29/discussion/#comment-1494821417'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

MKhalusova replied:
<pre class="ql-syntax" spellcheck="false">Line #3.    verification_question =  "Where was this coding language developed: {{entity}}?"
</pre><p>How do we choose the verification question?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/29/discussion/#comment-1494821418'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

MKhalusova replied:
<pre class="ql-syntax" spellcheck="false">Line #2.    # TODO: Update regex if the format of the baseline response changes. (ex. not a numbered list)
</pre><p>Another TODO - unclear if it's needed here</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/29/discussion/#comment-1494821419'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

MKhalusova commented:
A really interesting prompting technique! Thanks for adding! I left some feedback on the notebook, mainly - let's switch to an open source model (I saw that Mistral and Llama are supported), and remove unnecessary TODOs. 

In order to publish the notebook, a few more steps are needed:

* Add the notebook to the `_toctree.yml`
* Add the notebook to the list of the latest notebooks in the `index.md`. As this is the most recent addition, put it at the top of the list.
* Right after the first header (notebook title), add yourself as an author, like this: _Authored by: [Your Name](https://huggingface.co/your_profile)_  Feel free to use either your Hugging Face profile, or GitHub profile, it's up to you which one to link.

Ankush-lastmile commented:
- switched model usage to Mixtral
- rebase to main
- removed Todos
- added added descriptions

Ankush-lastmile commented:
Hi [@MKhalusova](https://github.com/MKhalusova)! Thanks for the review and the detailed comments. I've gone ahead and addressed your comments, as well as made the necessary changes you've requested. It would be great if you could take another look.

Thanks!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/cookbook/pr_29). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

MKhalusova commented on notebooks/en/chain_of_verification.ipynb:
```diff
@@ -0,0 +1,490 @@
+{
```
<p>Minor nit: Git<strong>H</strong>ub (H should be capitalized)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/29/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

MKhalusova replied:
<p>This paragraph can be much shorter: </p><p> </p><p> </p><p>The next code cell sets up a 'CoVe prompt template' within a structure known as AIConfig. AIConfig is a data format to organize prompt templates and specific model settings. To create a config you can either:</p><p>* Generate it using the <a href="https://marketplace.visualstudio.com/items?itemName=lastmile-ai.vscode-aiconfig" target="_blank">AIConfig Vscode Extension</a>, or</p><p>* Download the config <a href="https://github.com/lastmile-ai/aiconfig/blob/main/cookbooks/Chain-of-Verification/cove_template_config.json" target="_blank">here</a> and load it with: </p><p><code>config = AIConfigRuntime.load('cove_template_config.json')</code>. </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/29/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

MKhalusova replied:
<pre class="ql-syntax" spellcheck="false">Line #7.          "baseline_prompt": "Name 20 programming languages that were developed in the United States. Include the developer name in parantheses.",
</pre><p>minor, typo: it should be "parentheses"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/29/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

MKhalusova replied:
<pre class="ql-syntax" spellcheck="false">Line #1.    # verification_question = "Where was {{entity}} born?"
</pre><p>I think you can remove this comment</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/29/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

MKhalusova replied:
<pre class="ql-syntax" spellcheck="false">Line #10.        verification_completion = await config.run("verification", params, options=inference_options)
</pre><p>I'm getting an error on this line:</p><p> </p><pre class="ql-syntax" spellcheck="false">UnwrapError: Called <code>Result.unwrap()</code> on an <code>Err</code> value: "Missing API key 'OPENAI_API_KEY' in environment"
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/29/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

MKhalusova replied:
<pre class="ql-syntax" spellcheck="false">Line #5.    # Display with Markdown
</pre><p>There seems to be an issue rendering this output. Let's switch to a simple print. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/29/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

MKhalusova commented:
Thank you for iterating on the cookbook! I found a few typos and minor issues, but there's also an important one - the notebook throws an error:
 `UnwrapError: Called `Result.unwrap()` on an `Err` value: "Missing API key 'OPENAI_API_KEY' in environment"`

 Let's make sure it runs smoothly without errors.
